cpp/src/arrow/compute/kernels/scalar_compare.cc

```diff
@@ -68,6 +72,50 @@ struct GreaterEqual {
   }
 };
 
+struct BetweenLessEqualLessEqual {
+  template <typename T, typename Arg0, typename Arg1, typename Arg2>
+  static constexpr T Call(KernelContext*, const Arg0& middle, const Arg1& left,
+                          const Arg2& right, Status*) {
+    static_assert(std::is_same<T, bool>::value && std::is_same<Arg0, Arg1>::value &&
+                      std::is_same<Arg1, Arg2>::value,
+                  "");
+    return (left <= middle) && (middle <= right);
+  }
+};
+
+struct BetweenLessThanLessEqual {
+  template <typename T, typename Arg0, typename Arg1, typename Arg2>
+  static constexpr T Call(KernelContext*, const Arg0& middle, const Arg1& left,
+                          const Arg2& right, Status*) {
+    static_assert(std::is_same<T, bool>::value && std::is_same<Arg0, Arg1>::value &&
+                      std::is_same<Arg1, Arg2>::value,
+                  "");
+    return (left < middle) && (middle <= right);
+  }
+};
+
+struct BetweenLessEqualLessThan {
+  template <typename T, typename Arg0, typename Arg1, typename Arg2>
+  static constexpr T Call(KernelContext*, const Arg0& middle, const Arg1& left,
+                          const Arg2& right, Status*) {
+    static_assert(std::is_same<T, bool>::value && std::is_same<Arg0, Arg1>::value &&
+                      std::is_same<Arg1, Arg2>::value,
+                  "");
+    return (left <= middle) && (middle < right);
+  }
+};
+
+struct BetweenLessThanLessThan {
+  template <typename T, typename Arg0, typename Arg1, typename Arg2>
+  static constexpr T Call(KernelContext*, const Arg0& middle, const Arg1& left,
+                          const Arg2& right, Status*) {
+    static_assert(std::is_same<T, bool>::value && std::is_same<Arg0, Arg1>::value &&
+                      std::is_same<Arg1, Arg2>::value,
+                  "");
+    return (left < middle) && (middle < right);
+  }
+};
```

Comment:
I suggest to have a single `between` function that uses `FunctionOptions` to control its behavior. These different comparisons can be represented using an enum, which can then be used as a template parameter for compiler to dispatch accordingly.  See for example how [`round` function](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc#L901) is implemented.

Reply:
@edponce Thanks for the suggestion.

Reply:
@edponce Would function options also be better to use for the compare function?

Reply:
Thanks.

Reply:
Both Compare and Between will need a further option for strings similar to [SortKeys](https://arrow.apache.org/docs/cpp/api/compute.html#_CPPv4N5arrow7compute11SortOptionsE)

Reply:
Having between as a kind of compare function makes sense as they have similar code structures. Giving the option to access between or the between functions directly seems nice if one wants to keep the same structure as the compare functions.

Reply:
I don't think we want to keep the structure of the compare functions though. CompareOptions is going to be removed (ARROW-13378) and CompareFunction is really only there to make tests more convenient. I don't think we need to make between act like the compare functions, and I still agree with Eduardo that a structure like the round function makes more sense.

Reply:
Also because the current structure gets exposed in Python oddly (as four separate functions).

I think either we should have one function "between" with BetweenOptions, or we should have four separate functions with no options, but the current state where we have both options and four separate functions doesn't make sense.

Reply:
What was the reason for deprecating options from the compare function? Will the same apply to between?
It seems possible to call other functions with variants, for example log10 and log10_checked. A smaller API is nice since it means easier binding creation and maintenance in other programming languages, though still want user friendly interfaces.

Reply:
The CompareOptions weren't really options, it was just a facade in C++ over the underlying functions (somewhat similar to ArithmeticOptions in fact). When the options classes were "upgraded" to make them serializable, CompareOptions was deemed redundant and mostly removed.

Looking at it now, I can see the argument for four separate functions for consistency, but that means we probably don't want/need BetweenOptions or any enum at all.

Reply:
Ok. Am happy to attempt something similar to the round function, though it would be nice if it is for both between and compare functions since that would make maintenance and code reuse easier.

For string comparison, some options that give a comparison key similar to a sort key will likely be needed.

@edponce Light review would be helpful.

Reply:
Sorry - why do we need a sort key for string comparisons?

Reply:
Strings are sorted by iso codes. The codes may not be the appropriate ordering in all cases, see for example, - https://en.wikipedia.org/wiki/Alphabetical_order#Language-specific_conventions
- https://www.postgresql.org/docs/current/collation.html
- https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/xtfrm

Reply:
AFAIK, Arrow string transformations operate on ASCII and UTF-8 encodings. Support for any other encodings would need to be discussed further as its implications are not trivial wrt to implementation.

Reply:
Ok. Even with UTF-8, sorting and comparison is not trivial, but a FunctionOption for that makes sense. This can be implemented later. 

@kou @pitrou At present having four separate between functions seems reasonable. Will add Python bindings. Ideally, this could end up in 7.0.0 - then can request feedback from users and either keep the current structure for Between and Compare or use correctly implemented FunctionOptions for both Between and Compare, rather than having a wrapper - this would end up in 8.0.0.

Reply:
I also like the option (1) in https://github.com/apache/arrow/pull/11882#issuecomment-1002956781 . (One between function with options.)